### PR TITLE
Add basic docs for the parser crate

### PR DESCRIPTION
**Pull request by dhruvmanila**
## Summary

This PR adds a basic README for the `ruff_python_parser` crate and updates the CONTRIBUTING docs with the fuzzer and benchmark section.

Additionally, it also updates some inline documentation within the parser crate and splits the `parse_program` function into `parse_single_expression` and `parse_module` which will be called by matching against the `Mode`.

This PR doesn't go into too much internal detail around the parser logic due to the following reasons:
1. Where should the docs go? Should it be as a module docs in `lib.rs` or in README?
2. The parser is still evolving and could include a lot of refactors with the future work (feedback loop and improved error recovery and resilience)

### CODE REVIEW

File: crates/ruff_python_parser/src/lib.rs [hunk 1]
```diff
@@ -156,6 +159,28 @@ pub fn parse_program(source: &str) -> Result<ModModule, ParseError> {
     }
 }
 
+/// Parse a full Python program into a [Suite].
```
Comment (AlexWaygood) [inline crates/ruff_python_parser/src/lib.rs:162]:
```suggestion
/// Parse a full Python program into a [`Suite`].
```

File: crates/ruff_python_parser/src/parser/mod.rs [hunk 2]
```diff
@@ -124,54 +129,73 @@ impl<'src> Parser<'src> {
...
-                    ParseErrorType::UnexpectedExpressionToken,
-                    self.current_token_range(),
-                );
+    /// Parses a single expression.
+    ///
+    /// This is to be used for [`Mode::Expression`].
+    ///
+    /// ## Recovery
+    ///
+    /// After parsing a single expression, an error is reported and all of the remaining tokens are
```
Comment (AlexWaygood) [inline crates/ruff_python_parser/src/parser/mod.rs:151]:
```suggestion
    /// After parsing a single expression, an error is reported and all remaining tokens are
```

File: crates/ruff_python_parser/src/parser/mod.rs [hunk 3]
```diff
@@ -124,54 +129,73 @@ impl<'src> Parser<'src> {
...
+    /// dropped by the parser.
+    fn parse_single_expression(&mut self) -> ast::ModExpression {
+        let start = self.node_start();
+        let parsed_expr = self.parse_expression_list(ExpressionContext::default());
+
+        // All of the remaining newlines are actually going to be non-logical newlines.
```
Comment (AlexWaygood) [inline crates/ruff_python_parser/src/parser/mod.rs:157]:
```suggestion
        // All remaining newlines are actually going to be non-logical newlines.
```

File: crates/ruff_python_parser/src/token.rs [hunk 4]
```diff
@@ -352,6 +352,10 @@ impl fmt::Display for Tok {
     }
 }
 
+/// A kind of token.
+///
+/// This is a lightweight representation of [Tok] which doesn't contain any information
```
Comment (AlexWaygood) [inline crates/ruff_python_parser/src/token.rs:357]:
```suggestion
/// This is a lightweight representation of [`Tok`] which doesn't contain any information
```

**Review by AlexWaygood — APPROVED** [review]

**Comment by dhruvmanila** [comment]
Hmm, I think I can just use the following in `lib.rs`:
```rs
#![doc = include_str!("../README.md")]
```

**Review by MichaReiser — APPROVED** [review]

**Pull request merged**